### ISSUE TITLE
DataStore Query Performance Very Slow

### ISSUE THREAD

**Issue by sacrampton** [aws-amplify/amplify-js#8405]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React Native

### Amplify APIs

DataStore

### Amplify Categories

api

### Environment information

<details>

```
# Put output below this line


```

</details>

### Describe the bug

We are porting our AppSync/GraphQL based app to DataStore and what we are finding is that queries on the cached database are very slow. We are also finding that pagination of the queries makes no difference to the performance of the queries.

There are many tables with the largest table having about 8,000 records and most other tables having less than 1000 records. There are only 5 tables/models that have data.

Whether we are connected to the network or working offline it makes no difference to the performance.

In issue #6994 it talks of a first time slowing in performance, but we see consistently slow performance.

I have recorded a video of the performance we are seeing - both online and offline - to give you an idea of the problem we are facing.

https://user-images.githubusercontent.com/6362888/120909860-f74b7180-c6bc-11eb-97dc-300f8d19ab4d.mp4

### Expected behavior

The database is local so we should be seeing almost instant responses to our queries. The performance of our existing AppSync GraphQL cache gives us almost instant responses so we expect the same or better performance than the existing AppSync GraphQL database.

### Reproduction steps

Not sure how to replicate this as the database I have has 48 models and 200 GSI's and takes about 6 hours of incremental (manual) pushes to deploy into a new environment.

### Code Snippet

```javascript
// Put your code below this line.

```

### Log output

<details>

```
// Put your logs below this line


```

</details>

### Mobile Device

iPhone 7

### Mobile Operating System

14.5

### Mobile Browser

Not relevant

**Comment by nubpro** [comment]
I closed my original issue as I could no longer replicate the problem anymore

Do u happen to be using the Storage module in your app?

**Comment by sacrampton** [comment]
Hi @nubpro - we are using the Storage module in the app for upload of photographs.

**Comment by sacrampton** [comment]
New Video - here is exactly the same application - with a few hundred records loaded versus over 7,000 in the previous video.

Seems loading more than a trivial amount of data is grinding it to a halt.

https://user-images.githubusercontent.com/6362888/120977328-8e482480-c7b6-11eb-9c10-61daf9a3076a.mp4

**Comment by sacrampton** [comment]
Hi @manueliglesias  / @iartemiev 

Is there anything I can provide you to move this along? Its on a critical time line for us to get this resolved.

Thanks

Stephen

**Comment by iartemiev** [comment]
@sacrampton is the performance comparable between Simulator and physical devices on both iOS and Android? Or is one worse than the other? Does query performance degrade gradually as the number of records increases? Or does it tend to be acceptable up to a certain point and then rapidly drops? 

Since you said only 5 of your models have data in them, I'll attempt to reproduce this with 5 models and 10k+ records in each. Will let you know what I find.

**Comment by iartemiev** [comment]
Also, which version of aws-amplify are you currently using?

**Comment by sacrampton** [comment]
Hi @iartemiev 

I went through the list of models and data counts in more detail. The difference between the data in each model between the first and second videos is shown here.

Thanks...

<meta name="ProgId" content="Excel.Sheet">
<meta name="Generator" content="Microsoft Excel 15">
Data By Model

</style>





Name | High Count | Low Count
-- | -- | --
Asset | 7,409 | 359
AssetVisit | 1,127 | 100
Photo | 4,667 | 352
ClientPickList | 1,233 | 1,233
FailureMode | 25 | 25
FieldList | 2,540 | 2,540
FieldSet | 147 | 147
GlobalPickList | 2,970 | 2,970
OperatingUnit | 1 | 1
Package | 34 | 34
PhotoButton | 13 | 13
Plant | 1 | 1
Risk | 37 | 37
Scale | 53 | 53
Uom | 1,099 | 1,099
User | 1 | 1
&nbsp; | 21,357 | 8,965

**Comment by sacrampton** [comment]
Hi @iartemiev - here are a few other things I'd like to report - I figure if we solve any of them they are all related and that will unravel the rest of them.

- We do the base query to hydrate DataStore on the mobile - the queries you are seeing in the attached videos are on the data that has been downloaded with the base queries.
- If I enter data in a field on the mobile and save that I see the update on the server straight away - but I don't see the data I entered on the mobile after I entered it. I mean if I go back into that record all I see is what was there when I did the base query (the data that was entered in the mobile app is not there)
- Eventually, after 10-15 minutes the data I entered is visible on the mobile.
- If I enter data on the server I never see the subscription coming across unless I do the base query refresh again

I figure all these issues are all inter-related - but the fact they all sort of work after a period of time (except the subscription) is making it hard to work out what is going on.

**Comment by sacrampton** [comment]
Using aws-amplify version = 3.3.27

**Comment by sacrampton** [comment]
Hi @iartemiev - I had another thought on this. On the Asset object I have potentially 400+ attributes, and potentially similar on the AssetVisit object. In DynamoDB there would rarely be more than 30 attributes on any individual record so the size is quite small. And in GraphQL we just download the attributes we need for a specific app so we are never trying to download 400+ attributes.

But I'm starting to think that DataStore might is creating 400+ attributes on every single record on the mobile. In my case they would be hundreds of empty attributes taking up space. Is that what might be going on? Can you test by creating a couple of models with hundreds of empty attributes and see if that is the issue?

**Comment by iartemiev** [comment]
Thanks for the additional info. Are you returning all of the records from when you're querying Asset, AssetVisit, etc.? Or are you using predicates to return a subset?

**Comment by sacrampton** [comment]
Hi @iartemiev - sorry I'm probably not understanding the question. Our app is multi-tenanted. So we have a base query to return only the data that belongs to that client. But in our existing GraphQL app we would only return the specific attributes from say the Asset that the App needs (say 30 attributes and not 400+ attributes).

Also, I notice in the GraphQL when it is a BELONGS_TO association that the GraphQL query is returning all the associated data and I guess it is storing that against the record. For example, and Asset in most cases has a ParentAsset - so it seems that not only is the Asset getting all 400+ blank attributes, it is storing the 400+ blank attributes of the ParentAsset. The Asset has a lot of HAS_MANY connections, so each of those objects is also getting the 400+ blank attributes of the connected Asset in its GraphQL query.

**Comment by sacrampton** [comment]
Thanks @iartemiev. I have a further question related objects - whether the data from the related object is embedded in the original object inside DataStore. The queries tend to suggest that might be the case.

So for example, I might have an optional connection to a parent Post
```
type Post @model {
  id: ID!
  parentPost: Post @connection (fields: "postParentPostId")
  postParentPostId: ID
  title: String!
  field1: String
  field2: String
  field3: String
  field4: String
  field5: String
}
```
I am wondering if the Parent record is embedded in the current record like something like this.
```
{
  "id": "id1",
  "title": "someTitle",
  "postParentPostId" : "id2",
  "field1": null,
  "field2": null,
  "field3": null,
  "field4": null,
  "field5": null
  "parentPost" {
     "id": "id2",
     "title": "someTitle",
     "postParentPostId" : "id3",
     "field1": null,
     "field2": null,
     "field3": null,
     "field4": null,
     "field5": null
      }
}
```

Also, editing the schema.js is fine - I can do that. Do I need to edit any of the other files (ie. index.d.ts, etc.)

And I guess I can have a different schema.js for each application I have so each application can get to just the attributes it needs.

**Comment by sacrampton** [comment]
Hi @iartemiev - to answer your question **_"...I meant when you're calling DataStore.query, are you passing it a predicate, e.g., DataStore.query(Asset, c => c.someField('eq', 'someValue') or perhaps using pagination or are you retrieving all of the records from the local store by calling DataStore.query(Asset) (no predicates or pagination)?..."_**

We are doing the base query with a predicate as you describe and we are using pagination (500 records per cycle) because 1000 causes a resource exceeded error.

**Comment by sacrampton** [comment]
Hi @iartemiev - if we expand your type Post model above and have 100 x "field01": null fields and 2 fields as you show. In DynamoDB you would just have the id/title fields and this would be 32b in size. But if I look at the size of adding 100 null fields it swells to 1,532b. On a mobile where we are trying to save memory then it seems counterintuitive to be storing null fields. Is there any opportunity to store like DynamoDB? I guess not since  I think you are using SQLite..

**Comment by sacrampton** [comment]
Hi @iartemiev - Some more results of testing with the original data set of 21,357 records detailed above. Querying the asset list before was taking 12 seconds. I have created a new schema.js that reduced the number of attributes on the records down significantly for the 3 models below.

&nbsp; | Original | New
-- | -- | --
Asset | 371 | 121
AssetVisit | 352 | 43
Photo | 112 | 60

The result was that by reducing the number of attributes on the records as shown above halves the response time (6 seconds to retrieve the assets list versus 12 seconds with the the previous one.

I think this shows that syncing null attributes to a record can dramatically impact performance. Given that there is a direct connection between storing null attributes and performance it follows that if we can get rid of all null attributes (like DynamoDB does) then we are going to be maximizing performance. The only way I can think of to get rid of all null values at all times is if you were able to use some sort of a NoSQL repository on the device that only stores non-null attributes, thus optimizing space and performance?

**Comment by sacrampton** [comment]
Hi @iartemiev - another test result to share with you. Used another larger dataset for testing. This data set had these quantities in these 3 models. All the other models are approximately the same quantities - they don't vary by project as they are essentially reference data objects.

&nbsp; | Records
-- | --
Asset | 30,998
AssetVisit | 9,315
Photo | 25,776

I tried with the Original schema.js and the New schema.js defined above.

- Original schema took approximately 10 minutes to download to the cache, then an attempt to query the Asset model results in the app crashing with memory issues.
- New schema took approximately 8 minutes to download to the cache, but then the Asset model was successfully queried in 1 minute and 4 seconds.

Note that when we query a model the response is irrelevant to the amount of data we are retrieving. From the 30K assets I could do a query that retrieves 1 record and it will take 1 minute and 4 seconds. I could conduct another query that returns 5K records and it take 1 minute and 4 seconds.

**Comment by sacrampton** [comment]
Hi @iartemiev - one final update from my testing. On the desktop you ship a storage adapter for IndexedDB which is a NoSQL database. So we did a test on the desktop with the large 30K+ Asset data set described above using the full schema.

It took about 11 minutes for the full dataset to download - but after that a query of the Asset table too **_2 seconds_**. This compares with 64 seconds on SQLite.

**Comment by iartemiev** [comment]
I'll try to respond to the questions/points in order

> I have a further question related objects - whether the data from the related object is embedded in the original object inside DataStore. 

For Has One and Belongs To relationships we resolve the connection at run time when you call `DataStore.query` on that model, but they are not saved that way to the local store.

> We are doing the base query with a predicate as you describe and we are using pagination (500 records per cycle) because 1000 causes a resource exceeded error.

By "base query" are you referring to the initial base sync query between AppSync and the local store? Or calling `DataStore.query` to retrieve records from the local store?

> On a mobile where we are trying to save memory then it seems counterintuitive to be storing null fields. Is there any opportunity to store like DynamoDB? I guess not since I think you are using SQLite..

That's not possible at the moment, but it's certainly something we will explore in the future. 
On React Native, DataStore utilizes [Async Storage](https://react-native-async-storage.github.io/async-storage/) for the local store implementation. Async Storage in turn uses different underlying persistence layers depending on the platform. On iOS it uses a serialized dictionary/key-value store, not SQLite.

> Note that when we query a model the response is irrelevant to the amount of data we are retrieving. From the 30K assets I could do a query that retrieves 1 record and it will take 1 minute and 4 seconds. I could conduct another query that returns 113 records and it take 1 minute and 4 seconds.

Could you please share the query you're using here? Specifically, what predicate are you passing?

**Comment by sacrampton** [comment]
Hi @iartemiev - thanks for answering my long chain of rambling "input" / observations.

- "base query" - https://docs.amplify.aws/lib/datastore/how-it-works/q/platform/js#sync-data-to-cloud ("...The Sync Engine will run a GraphQL query on first start that hydrates the Storage Engine from the network using a Base Query...")
- Async Storage - thanks for that - we are using React Native on mobile and Angular on browser.
- Model and Query shown below.

```
type Asset @model
  @key(name: "byAssetPackageId", fields: ["assetPackageId"]) 
  @key(name: "byAssetPlantId", fields: ["assetPlantId"]) 
{
  id: ID!
  name: String!
  plant: Plant @connection (fields: ["assetPlantId"])
  assetPlantId: ID!
  package: Package @connection (fields: ["assetPackageId"])
  assetPackageId: ID
  auditCompleted: Boolean
}

type Plant @model
{
  id: ID!
  name: String!   
}

type Package @model
{
  id: ID!
  name: String!   
}
```

The base query would be to download all assets in a Plant
```
query baseQuery{
  syncAssets(filter:{
    assetPlantId:{eq:"xxxxxxxxxxxx"}
  }){items{id _version}}
}
```

Then in DataStore with the 30K assets we do a query to find Assets in a package that have not yet been audited. This query takes 1 minute 4 seconds to return the 2 records. If we remove the limit it returns 113 records and takes the same time.
```
const data = await DataStore.query(
 Asset,
 (asset) =>
 asset
 .assetPackageId('eq', userData.package.id)
 .auditCompleted('eq', false),
 {
 page: 0,
 limit: 2,
 },
 );
```

**Comment by iartemiev** [comment]
Thank you, that is helpful. We'll be doing additional library benchmarking today in order to pinpoint the issue and will provide an update with our findings.

**Comment by iartemiev** [comment]
Quick update: I have been able to repro performance issues similar to those described above when using a model with 7 fields containing values + 100 fields with `null` values and > 30k records in the local store. Will be investigating causes and potential solutions.

**Comment by sacrampton** [comment]
Hi @iartemiev - that is fantastic news that you were able to reproduce. Hopefully you can also reproduce it working perfectly with IndexedDB/Browser too. Await your feedback.

**Comment by iartemiev** [comment]
We've pinpointed the bottleneck to the `multiGet` [method](https://react-native-async-storage.github.io/async-storage/docs/api#multiget) in Async Storage, it seems that we're brushing up against that library's performance limitations here.

We'll be brainstorming what improvements we could make in the library in order to mitigate this as much as possible. For your use case specifically, even with the reduced number of fields, are a considerable number of values in your records still `null`? If so, we have a couple of ideas on how we may be able to optimize for this.

As an aside, when you're seeing this poor performance do you have the React Native debugger enabled on your app? We're seeing significantly better performance (still suboptimal, but 2-3x faster) with the debugger disabled, so just want to point that out as well.

**Comment by sacrampton** [comment]
@iartemiev - looks like good progress.

Ultimately, I'd like to retain access to all fields (like we do with IndexedDB) - but for now we can run with the reduced schema. And even in the reduced schema I would estimate that 80% of those fields are null in the dataset.

I don't believe we have debugger enabled - but I will check.

**Comment by sacrampton** [comment]
@iartemiev - checked performance with and without debugger enabled. The 64 second response was with React Native Debugger enabled. With Debugger switched off the response is 22 seconds. As you said - its about a 3X improvement.

Worth remembering that 22 seconds was with the stripped down schema and is still 11X slower than the Browser/IndexedDB running with the full schema. Good to know we'll see that sort of improvement when we go to production, but 22 seconds with cut-down schema still isn't a solution.

Thanks for working this.

**Comment by iartemiev** [comment]
To level-set on expectations a bit, I don't think we'll be able to get the same performance out of AsyncStorage as we do from IndexedDB, no matter what optimizations we add. The former is an intrinsically slower persistence layer when dealing with large amounts of data. I'm hoping we can see another 2-3x improvement, but 10x is quite unlikely. In order to achieve an order of magnitude improvement, we'll need to leverage a faster local database for React Native. We'll certainly explore that option, but it's not something we'd be able to release in a matter of a day or two.

**Comment by sacrampton** [comment]
Hi @iartemiev - just checking back in to see whether you have made any progress on this.

**Comment by iartemiev** [comment]
Hey @sacrampton, we've identified the optimization we're planning on implementing and it's what you suggested above, i.e., not persisting `null` values to the local databases. We're testing this solution now to make sure that it doesn't have any unwanted side effects.

If you would like to test the effect this change will have on your queries before we release the fix, you can make the following **_temporary_** change in your Amplify dependencies manually:

**Note you'll want to delete your app and re-install it, i.e., re-sync the local database in order for the change to take effect**

1. Open `./node_modules/@aws-amplify/datastore/lib-esm/datastore/datastore.js`
2. Change this code (should start at line 279)
```js
var initializeInstance = function (init, modelDefinition, draft) {
    var modelValidator = validateModelFields(modelDefinition);
    Object.entries(init).forEach(function (_a) {
        var _b = __read(_a, 2), k = _b[0], v = _b[1];
        modelValidator(k, v);
        draft[k] = v;
    });
};
```
To:
```js
var initializeInstance = function (init, modelDefinition, draft) {
    var modelValidator = validateModelFields(modelDefinition);
    Object.entries(init).forEach(function (_a) {
        var _b = __read(_a, 2), k = _b[0], v = _b[1];
        modelValidator(k, v);
        if (v !== null) {
            draft[k] = v;
        }
    });
};
```

3. If your app happens to be using CJS modules, instead of ESM, you'll want to change the same code in `./node_modules/@aws-amplify/datastore/lib/datastore/datastore.js` instead (`lib` subdirectory as opposed to `lib-esm`)

**Comment by sacrampton** [comment]
Hi @iartemiev - that sounds great - will get testing immediately and let you know how we go.

I just created another issue (#8450) also dealing with null values and subscriptions. Will the changes you made here deal with that?

**Comment by sacrampton** [comment]
Hi @iartemiev - we did the testing and can indeed see that the null values are removed from the data. However, it didn't change the performance we were seeing. Namely 64 seconds in debug and 24 seconds without debug.

**Comment by iartemiev** [comment]
Hmm that's odd, as we're seeing a roughly 2-2.5x improvement with the above change. Did you reinstall the app and re-sync the records after changing the code? 

If so, could you please verify that the fields containing `null` values are definitely no longer getting persisted to AsyncStorage?

You can do that by querying AsyncStorage directly (without DataStore):

```js
import AsyncStorage from '@react-native-async-storage/async-storage';

async function getFromAsyncStorage() {
  const model = 'Asset';

  const key = (await AsyncStorage.getAllKeys()).find(key =>
      key.includes(`@AmplifyDatastore::user_${model}`)
  );

  console.log(await AsyncStorage.getItem(key));
}
```

**Comment by sacrampton** [comment]
Hi @iartemiev - will check this out and confirm back with you asap.

If you have a chance, can you comment on this somewhat related issue of nulls in DataStore (#8450)

**Comment by sacrampton** [comment]
Hi @iartemiev - have double checked everything you said and the result is still the same.

Here is a screen shot of the data we see on the record after applying your fix - all the null values are removed.
![newDataStore](https://user-images.githubusercontent.com/6362888/122216512-d28e9f80-ceef-11eb-87d5-fb3394fceee7.png)

When we try the code you list above we are getting this - JSON value '<null>' of type NSNull cannot be converted to NSString

[AsyncStorage] Using undefined type for key is not supported. This can lead to unexpected behavior/errors. Use string instead.
Key passed: undefined

**Comment by iartemiev** [comment]
If the screenshot above is the result of a `DataStore.query`, that doesn't tell us definitively whether the `null`s are being persisted or not to AsyncStorage, just that they're being omitted when the model class is instantiated by DataStore.

If the code I provided didn't work, you can log out all the keys with `console.log(await AsyncStorage.getAllKeys())`, then pass one of those keys to `await AsyncStorage.getItem("some-key")` and log that to the console. 

It may be easier to set AsyncStorage to a global and then interact with it in real time via the RN debugger console.

In your app:
```js
import AsyncStorage from '@react-native-async-storage/async-storage';
global.AsyncStorage = AsyncStorage;
```

Then in the debug console you can run e.g., 
```js
await AsyncStorage.getAllKeys();

// pick one of the keys that starts with "@AmplifyDatastore::user_{your model}"
// call getItem on that key

await AsyncStorage.getItem("@AmplifyDatastore::user_...")
```

**Comment by sacrampton** [comment]
Hi @iartemiev - the team has been working to get this done, but not getting much progress.
<img width="730" alt="MicrosoftTeams-image (1)" src="https://user-images.githubusercontent.com/6362888/122394418-4e085380-cfb9-11eb-8321-0c46d44cdb4a.png">

"I am only getting keys in which i store the data using - awaitAsyncStorage.getAllKeys()"

"I am not getting any key like - @AmplifyDatastore::user_Asset"

That was checked in the React Native debugger console.

**Comment by iartemiev** [comment]
Ah, I just realized that you're on an older version of aws-amplify that still uses the AsyncStorage module that's bundled with React Native, as opposed to the external module. You should be able to query it by import from RN:
```js
import { AsyncStorage } from 'react-native'
```
You should then see the DataStore-specific values in there using the code I provided above

**Comment by iartemiev** [comment]
(Apologies, finger slipped and I accidentally clicked "Close with comment" instead of "Comment". Re-opened the issue)

**Comment by sacrampton** [comment]
Hi @iartemiev - the team has checked using await AsyncStorage.getItem(key).

The code used before looks like this
![codeBefore](https://user-images.githubusercontent.com/6362888/122477212-b9c6dc80-d00a-11eb-883c-0d8e29cac49f.png)

This results that come are shown below
<img width="1435" alt="resultBefore" src="https://user-images.githubusercontent.com/6362888/122477272-cba87f80-d00a-11eb-99f8-ca544ed5c21c.png">

After applying null validation we are getting these field in asset record. Here is the code after.
![codeAfter](https://user-images.githubusercontent.com/6362888/122477351-e67af400-d00a-11eb-9535-dbaa5831b33b.png)

And this is the data we are seeing
<img width="1133" alt="resultAfter" src="https://user-images.githubusercontent.com/6362888/122477454-14603880-d00b-11eb-8a77-e7879a9c18cf.png">

So it is clear that the null fields are gone after updating the code in aws-amplify module - but the speed doesn't change.

**Comment by sacrampton** [comment]
Hi @iartemiev - to rule out any related issues we upgraded aws-amplify to 4.1.1 - still same performance issue

**Comment by sacrampton** [comment]
Hi @iartemiev - just back tracking for a second. I significantly reduced number of attributes on 3 tables and assumed it was removal of all those null values which made the difference. Maybe it was the fact of reducing attributes on a table?

**Comment by iartemiev** [comment]
Reducing the amount of data stored in AsyncStorage will increase query performance. Whether that’s done through explicitly removing fields or by not persisting null values, the effect should be roughly the same.

**Comment by iartemiev** [comment]
Hey @sacrampton, wanted to share an update with you: we've decided to implement a new storage adapter option for React Native CLI apps that utilizes SQLite as the local persistence layer. We saw massive performance improvements when querying with predicates, sorting, and/or pagination, as well as processing the base/delta sync queries compared to the performance of AsyncStorage. 

We still have some testing to do before we're ready to release this, but it's about 90% implementation complete as of right now. Once we're satisfied with the stability of this adapter, we'll let you know how you can utilize it in your app. 

As an aside, you mentioned above that you're on version 3.3.27 of Amplify. Is there something in particular that's keeping you from upgrading to the latest version?

**Comment by sacrampton** [comment]
That is awesome @iartemiev - I had the guys upgrade to 4.1.1 - there was nothing keeping us from upgrading. Look forward to hearing more.

**Comment by sacrampton** [comment]
Hi @iartemiev - sounds great. Cascading deletes don't matter to me at this time.

**Comment by iartemiev** [comment]
Update: ran into a hiccup when testing many-to-many relations. Got a solution in mind, but will need just a bit more time to implement it and test. I will provide another update tomorrow

**Comment by sacrampton** [comment]
Hi @iartemiev - appreciate you keeping me updated. Thanks.

**Comment by iartemiev** [comment]
Hello again, @sacrampton! We’ve released a special version of AmplifyJS (**aws-amplify@rn-sqlite**) containing the new SQLite adapter. We’re still considering this to be an experimental feature, hence the special tag/version.

We’ve tested this adapter thoroughly on our end with up to 350k records. Querying on a predicate that returns a subset of data or using pagination and/or sorting is blazing fast (a fraction of a second in our testing). If you want to return the entire contents of a large table (over 10k records), i.e., querying _without_ a predicate, that still takes a while (10+ seconds), but that is expected.

There is currently a bit of occasional funky behavior and console warnings when deleting relational records. This is because cascading deletes are not yet implemented. You would need to manually delete from the last descendant up to avoid warnings. During my testing, these did not break the application. We will be adding cascading delete functionality shortly to bring this to feature parity with the other adapters.

Please let us know in this thread if you run into issues, and we will get them resolved ASAP. We will include this adapter in the stable release soon but wanted to give you a chance to test and benchmark early to ensure that your performance needs are met.

Here are the installation instructions:

(note: AsyncStorage is still a required dep, as other parts of Amplify continue to rely on it) 
```py
yarn add aws-amplify@rn-sqlite amazon-cognito-identity-js@rn-sqlite aws-amplify-react-native@rn-sqlite react-native-sqlite-storage @react-native-async-storage/async-storage

# to ensure we're not left with any of the previous versions
rm -rf node_modules yarn.lock package-lock.json
yarn

npx pod-install
```

In order to restore the current/async-storage version:
```py
yarn remove react-native-sqlite-storage
yarn add aws-amplify@latest amazon-cognito-identity-js@latest aws-amplify-react-native@latest

# to ensure we're not left with any of the previous versions
rm -rf node_modules yarn.lock package-lock.json
yarn

npx pod-install
```

**Comment by sacrampton** [comment]
Much appreciated @iartemiev - will start building asap and let you know our results. Thanks again

**Comment by sacrampton** [comment]
Hi @iartemiev - I want to outline precisely what we did and what we are seeing in case there is anything we overlooked.

These were the installation instructions that we followed.

```
yarn add aws-amplify@rn-sqlite amazon-cognito-identity-js@rn-sqlite aws-amplify-react-native@rn-sqlite react-native-sqlite-storage @react-native-async-storage/async-storage
# to ensure we're not left with any of the previous versions
rm -rf node_modules yarn.lock package-lock.json
yarn
npx pod-install
```

After the above installation the libraries installed are as shown here.

![libraries](https://user-images.githubusercontent.com/6362888/123729592-8f153780-d8d8-11eb-8a91-69aab4c517f4.png)

We then delete the old app and reinstall the app again.

We still see the same 22 seconds when querying on 30K asset records and 35 seconds for 42K asset records.

We are importing DataStore like this

```
import {DataStore} from 'aws-amplify';
```

Appreciate any thoughts you have on why we might not be seeing the same as you are seeing.

**Comment by sacrampton** [comment]
Hi @iartemiev - some more updates here.

We deleted node modules and pod files, and follow all the steps above, but when we check the code it is still getting data from AsyncStorage.

```
const model = 'Asset';
 
 const key = (await AsyncStorage.getAllKeys()).find((key) =>
 key.includes(`@AmplifyDatastore::user_${model}`),
 );
 
 console.log('Item from aync storage >>', await AsyncStorage.getItem(key));
```

So clearly we are not doing something right here. Can you see if there is anything obvious that we are omitting.

Suspect something is being cached in existing project so will create a new project from scratch, but that will take several hours, an see if that works for us.

**Comment by sacrampton** [comment]
Hi @iartemiev - getting errors with our schema. We have a modified schema.js that works fine with AsyncStorage when running Datastore.start()

error: no such table: Setting
Possible unhandle promise rejection
"code": 5, "message": "duplicate column name: assetOperatingUnitId",

So went back to the standard codegen model and got this error.


Error >> no such table: Setting backend.js:12506 Possible Unhandled Promise Rejection (id: 0): Object { "code": 5, "message": "near \"group\": syntax error", }

**Comment by iartemiev** [comment]
Ok, as of the last comment, it does look like you're now utilizing the SQLite adapter but it's having trouble generating the local SQL schema from your AppSync schema. 

Could you please share your entire `schema.graphql` file with me? If you don't feel comfortable with posting it here, you could add it to a private repo and then invite me as a collaborator.

**Comment by sacrampton** [comment]
Hi @iartemiev  - I can't share it here - but @manueliglesias has a copy of it he could provide to you.

**Comment by iartemiev** [comment]
@sacrampton I was able to test with your schema and identified the issue. We were not escaping column and table names when generating the SQLite schema, so reserved words in model fields names (such as `group`) were causing errors. Will be fixed in this PR: https://github.com/aws-amplify/amplify-js/pull/8519. We will publish a new version of **aws-amplify@rn-sqlite** once the PR is reviewed and merged

**Comment by iartemiev** [comment]
@sacrampton we just published an update: [aws-amplify@4.1.3-rn-sqlite.3](https://www.npmjs.com/package/aws-amplify/v/4.1.3-rn-sqlite.3). It fixes the error you're seeing (I validated it against your schema).

**Comment by sacrampton** [comment]
That's awesome @iartemiev  - will start testing immediately.

**Comment by sacrampton** [comment]
Hi @iartemiev - we were able to sync our model, but running into an issue with Boolean fields when querying.

We query the model and get this error - the schema for hasPhotos is Boolean. I did a scan of the DynamoDB database and there are no "rouge" non-boolean values for this attribute. Not sure what is going on with this.

```
     Error: Field hasPhotos should be of type boolean, number received. 0
```

**Comment by sacrampton** [comment]
Hi @iartemiev - just a question on _deleted fields. Noticed by default this is not added to the schema so we can't query to not return deleted records - or does DataStore automatically exclude those from queries?

**Comment by iartemiev** [comment]
We just fixed the Boolean error in  [aws-amplify@4.1.3-rn-sqlite.4](https://www.npmjs.com/package/aws-amplify/v/4.1.3-rn-sqlite.4). Please upgrade to that and let us know if you run into any other issues.

> just a question on _deleted fields. Noticed by default this is not added to the schema so we can't query to not return deleted records - or does DataStore automatically exclude those from queries?

Records with `_deleted: true` get deleted from the local store and should not be returned in the `DataStore.query` results.

**Comment by sacrampton** [comment]
Hi @iartemiev - After upgrading we are again getting this error which was previously seen.... Appreciate any ideas.

 ```   
Error during getting data from async storage. {​​​​​​​message: "no such table: Setting", code: 5}​​​​​​​
backend.js:12506 Possible Unhandled Promise Rejection (id: 0):
Object {​​​​​​​
  "code": 5,
  "message": "duplicate column name: assetOperatingUnitId",
}​​​​​​​
```

**Comment by iartemiev** [comment]
The only thing we've changed in the latest version is handling the conversion of a SQLite boolean (`0`, `1`) to a JS representation (`false`, `true`) everything else is exactly the same. 

Have you manually modified your schema.js since? A likely culprit here is a manually added foreign key, like the one in the error message above. 

When we generate the SQL schema, we have to account for every field that's going to be stored in the table beforehand and add it as a column. So we make inferences such as adding join table FKs automatically when we don't expect them to be added to the generated schema via codegen. This is not required in the NoSQL dbs, because they don't have fixed schemas. So what's likely happening is when we're creating the SQL tables, your manually added keys are duplicates of what we auto-generate, hence the message. 

I can add some field name de-duping logic for this (i.e., only auto-generate the FK if it's not explicitly added to the schema.js fields for that model). But in the meantime, if you take those fields out of schema.js but leave them in the type definitions (index.d.ts) that should unblock you.

**Comment by iartemiev** [comment]
Actually, scratch what I said re: taking those fields out. I'm not sure that will help if you need to access them in your app. Let me see if I can get an update out to you.

**Comment by sacrampton** [comment]
Hi @iartemiev - think it is something to do with applying update to existing project. Earlier we said we had to recreate from scratch

**Comment by iartemiev** [comment]
I don’t think that’s the case here. The last update made no changes to the storage adapter itself, so it should be fine to apply it to an existing app. The cause is most likely what I described above.

**Comment by iartemiev** [comment]
I opened a PR that addresses that by the way, but given that it’s a US holiday weekend, it might not get reviewed until Tuesday

**Comment by iartemiev** [comment]
@sacrampton we just released a new version - [4.1.3-rn-sqlite.5](https://www.npmjs.com/package/aws-amplify/v/4.1.3-rn-sqlite.5). For this update, I do recommend deleting and re-installing the app, since there are changes to the schema generation code.

**Comment by sacrampton** [comment]
Thanks @iartemiev - will let you know how we go

**Comment by sacrampton** [comment]
Hi @iartemiev - good news - SQLite now working and previously with our full schema it wouldn't even load 31K assets - and with the massively trimmed down schema it was 64 seconds to query 31K assets. Now we can use the full schema and we are seeing 2 seconds to query 31K assets. Still testing, but looking really promising.

**Comment by iartemiev** [comment]
That's great to hear!

**Comment by iartemiev** [comment]
@sacrampton by the way, are you seeing 2 second query time with the RN Debugger enabled or disabled? I was getting records back in less than half a second with the debugger disabled even with hundreds of thousands of records in the model. But this may be due to hardware differences as well.

**Comment by sacrampton** [comment]
Hi @iartemiev - I can confirm the 2 second query is with RN Debugger disabled.

On Android we are running into issues with doing hydrating the cache where it is just getting stuck. See the log screen shot below.

<img width="1435" alt="MicrosoftTeams-image (4)" src="https://user-images.githubusercontent.com/6362888/124560296-fc4c3e00-de7f-11eb-8653-543e8e9c1e13.png">

Any ideas?

**Comment by iartemiev** [comment]
The warning you're seeing shows up when using the latest stable version of Amplify with the AsyncStorage adapter as well, it should not be a problem in and of itself. It's due to DataStore using a long `setTimeout` value  (300,000 ms) for ensuring we're still getting subscription keep-alive messages from AppSync. So unfortunately the screenshot doesn't give me any info on what could be going wrong. 

Follow up questions: 
* Are there any other debug messages, warnings, or errors in the console being emitted by Amplify or DataStore (particularly, after setting `Amplify.Logger.LOG_LEVEL = 'DEBUG'`?) If so, please either post them here or email them to me. 
* Was your app working on Android with the AsyncStorage adapter? 
* Does the sync always get stuck in the same place in the process (i.e., after a particular model is synced)?
* What happens when you try to query a model when the sync is stuck?

**Comment by sacrampton** [comment]
Hi @iartemiev - the sync stops then it does eventually start again, but it could take quite a while to restart. With the AsyncStorage adapter it had a limit of 6MB so our datasets were much smaller than those we are testing now. Getting more specific responses to your questions asap.

When we have synced all 31K assets in our test set the query is taking 5 seconds. This query previously on AsyncStorage on iOS was taking 64 seconds and now on SQLite on iOS taking about 2 seconds.

**Comment by sacrampton** [comment]
Hi @iartemiev - here are our more detailed answers.

_Are there any other debug messages, warnings, or errors in the console being emitted by Amplify or DataStore (particularly, after setting Amplify.Logger.LOG_LEVEL = 'DEBUG'?) If so, please either post them here or email them to me._
- I am getting the same waring multiple time specially before Asset and Photo models having large amount of data. I didn't getting any error this time using new schema even after setting upAmplify.Logger.LOG_LEVEL = 'DEBUG'?)

_Was your app working on Android with the AsyncStorage adapter?_
- No I am not able to use AsyncStorage Adapter I am getting this error even after increasing the storage size i.e. Error storing data Error: database or disk is full

_Does the sync always get stuck in the same place in the process (i.e., after a particular model is synced)?_
- It didn't get stuck but takes a long time to sync data using full schema.

_What happens when you try to query a model when the sync is stuck?_
- Never return if model didn't sync and try to make query.

**Comment by sacrampton** [comment]
Hi @iartemiev - we have built a signed APK to test the Android version on device and now are seeing performance comparable to what we were seeing on iOS.

**Comment by iartemiev** [comment]
That's awesome! Do you still have a lot of testing to do for your app on both iOS and Android before you're comfortable with the SQLite adapter? Or are you getting pretty close?

**Comment by sacrampton** [comment]
Hi @iartemiev - the performance of the AsyncStorage adapter basically makes it unusable on anything but exceptionally small data sets. So we are already an order of magnitude better off using the SQLite adapter. We will be going forward with the SQLite adapter and keep testing over the next week or so - but looks pretty good with where we are at now.

By the way, getting TOTAL's is really important to us. For example, to work out how many assets are in a work package - we want to be able to just query the "COUNT" of those records that meet that query. Seems to be a standard function of SQLite, but not available in DataStore's API. Is there a simple way you can enable the total number of items to be returned (not the items - just the total) give this seems to be a standard function of SQLite?

**Comment by iartemiev** [comment]
Very glad to hear that! 

Re: retrieving total row count. I definitely see the value in having that functionality. But since it would require a public API change to DataStore (e.g., we'd need to add a new public method or extend an existing one), we likely wouldn't be able to get that implemented very quickly since we have a particular process we must follow when introducing such changes. 

There is a workaround that you could leverage immediately though. You can open the database in your app in read-only mode and then execute a query against the table.

It would look something like this:
```js
import SQLite from 'react-native-sqlite-storage';
SQLite.enablePromise(true);

const DB_NAME = 'AmplifyDatastore';

let db;
function openDB() {
  db = await SQLite.openDatabase({name: DB_NAME, readOnly: true})
}

function closeDB() {
  if (db) {
    db.close();
  }
}

function getModelRecordCount(modelName) {
  if (!db) {
    await openDB();
  }

  const statement = `SELECT COUNT(*) AS count FROM ?`;

  const [resultSet] = await db.executeSql(statement, [modelName]);

  const result =
    resultSet &&
    resultSet.rows &&
    resultSet.rows.length &&
    resultSet.rows.raw &&
    resultSet.rows.raw();

  const [first] = result;
  return first.count;
}
```

Then when you need to retrieve the total, you can do:

```js
const assetCount = await getModelRecordCount('Asset');
```

**Comment by iartemiev** [comment]
Btw, the above code is just a rough example. You may need to modify it a little bit to suit your needs, but it should be close.

**Comment by iartemiev** [comment]
I do want to emphasize the importance of opening the DB in **read-only mode**, otherwise, you may experience some unpredictable behavior with DataStore.

**Comment by sacrampton** [comment]
Hi @iartemiev - thanks so much for the suggestion - will follow this and let you know how we go.

**Comment by jeremy-white** [comment]
@iartemiev I wanted to share with you an error I've come across when trying this out on my project

```
Object {
    "code": 5,
    "message": "table BrokenExample has no column named owner",
}
```

I use @auth and see an instance of that error for every model I use it on except for one. The one model that seems to work is the only one that I define the ownerField as "owner" and also have "owner" defined explicitly in the model definition, so something like 

```
type WorkingExample @model
    @auth(rules: [
        { allow: owner, ownerField: "owner", operations: [create] }
    ]) {
    id: ID!
    owner: String
}
```

versus

```
type BrokenExample @model
    @auth(rules: [
        { allow: owner, operations: [create] }
    ]) {
    id: ID!
}
```

Seems that explicitly defining the owner field like in the working example would be a fine workaround . . . but just wanted to make sure it's a situation you're aware of.

**Comment by sacrampton** [comment]
Hi @iartemiev - using the SQLite Adapter on RN we are noting something with caching and querying.

If I create a record the query will not return that record until it has been synced with the server. Its created locally, but query doesn't get anything till it has been synced. What are we doing wrong here?

```
await DataStore.save(new Asset({...assetData}));
```

```
const data = await DataStore.query(
 Asset,
 (s) => {
 if (!downloadEntirePlant) {
 return s
 .assetParentAssetId('eq', mobileActiveLocation?.id)
 .assetPackageId('eq', userPackageId)
 .assetPlantId('eq', userPlantId)
 .isDeleted('ne', true);
 } else {
 s.assetParentAssetId('eq', mobileActiveLocation?.id)
 .assetPlantId('eq', userPlantId)
 .isDeleted('ne', true);
 }
 },
 {
 sort: (s) => s.name(SortDirection.ASCENDING),
 },
 );
```

**Comment by iartemiev** [comment]
@sacrampton my guess is that it's got something to do with the predicate you're crafting.

You can test this with:
```js
const newAsset = await DataStore.save(new Asset({...assetData}));

const data = await DataStore.query(Asset, newAsset.id)
```

If this returns the expected data pre-sync, then I would look into adjusting the fields and/or values in that predicate. If not, there may be a bug in DataStore, so just let me know if that's the case.

**Comment by iartemiev** [comment]
@jeremy-white, thank you for calling this out. We'll need to make sure to add `owner` to the SQL schema in cases where it's implicitly defined (i.e., omitted from the model fields).

**Comment by sacrampton** [comment]
Thanks @iartemiev  - thanks for the feedback - can confirm your suggestion does indeed return data before it is synced to the server. Just need to work out how we get a list back now.

**Comment by jeremy-white** [comment]
Just a couple more things I've hit 

```
Object {
  "code": 6,
  "message": "NOT NULL constraint failed: TestModel.items",
}
```

Hit this for HAS_MANY connections defined as 

```
    items: [Item!] @connection(keyName: "byTestModel", fields: ["id"])
```

as well as 
```
    items: [String!]
```

I've been able to get past this by changing `isRequired` to false in schema.js . . . this could also be my misunderstanding but I was under the impression that [String!] meant that the items in the array had to be defined but the attribute itself could be undefined, so I was only using ! to be extra cautious. I can certainly work around this one as well but just wanted to raise it as it works as I'd expect with AsyncStorage but fails against SQLite

I'm also seeing
```
Error: Field items should be of type string, object received.
```

For fields I define as AWSJSON and have values that are arbitrary JSON . . . I haven't been able to workaround this other than just removing those fields from schema.js so if you have a better workaround I would love to hear it.

I really appreciate your work on this @iartemiev as I think it will be a HUGE improvement!

**Comment by sacrampton** [comment]
Hi @iartemiev - does anything in our predicate above jump out at you as a possible issue that is not getting un-synced data? We know we can retrieve a single un-synced record whose id we know. But the search for a list of records that included synced and un-synced records is something we are still struggling with. Any hints you might have on what we might be doing wrong in the above predicate would be much appreciated.

We are going live with our App this weekend - we wouldn't have got here without your amazing support in this effort. Much appreciated.

**Comment by iartemiev** [comment]
@sacrampton, I don't think there's anything inherently wrong with it. It's hard to know for sure without being familiar with your app code and knowing what values are being passed and what data is expected to be present in the local store. DataStore doesn't use a cache. Every time you execute `DataStore.query`, it performs a SQL query against your local database.

If I were debugging this, I would check that each of the operands has the expected value by logging them out to the console. 

You can also try removing fields from your predicates one by one to determine which one is causing all the results from being filtered out. 

Try something like the following in the same place as where you're normally attempting to query the local records (and before the sync):

```js
const newAsset = await DataStore.save(new Asset({...assetData}));
const data = await DataStore.query(Asset, newAsset.id);

console.log(data.assetParentAssetId === mobileActiveLocation?.id, data.assetParentAssetId, mobileActiveLocation?.id);
console.log(data.assetPackageId === userPackageId, data.assetPackageId, userPackageId);
console.log(data.assetPlantId === userPlantId, data.assetPlantId, userPlantId);
console.log(data.isDeleted === true, data.isDeleted, true);
```

If you have a `false` in any of those 4 console logs, that should give you the problematic field. 

Lastly, if you have debug logs on, you should see the SQL SELECT statement that is generated by that predicate. Does it look correct?

**Comment by sacrampton** [comment]
Hi @iartemiev  - the count suggestion you had above works for smaller data sets in iOS (extremely slow with larger data sets - > 1 minute). But in Android we get the following error.

```
{​​​​​​​​"message":"no such column: false (code 1 SQLITE_ERROR): , while compiling: SELECT * FROM Asset WHERE auditCompleted = false","code":0}​​​​​​​​​​​​​​​
```

**Comment by iartemiev** [comment]
@sacrampton SQLite represents boolean values as `0` and `1`. So if you're specifying a literal `false` in your WHERE clause it attempts to look up a column with the name "false" as the error message suggests. 

You can either specify `WHERE auditCompleted = 0` or use a parameterized query. The latter will coerce a JS boolean to the SQLite representation.

re: performance, I'm surprised it's running that slowly. I'll try to reproduce on my end to confirm.

**Comment by sacrampton** [comment]
Hi @iartemiev - replacing "false" with 0 is not working for us... The error seems to be that the column we are calling is not there.

**Comment by iartemiev** [comment]
What's the error you get when you query `WHERE auditCompleted = 0`?

By the way, are you calling `SELECT COUNT(*) as count FROM Asset`? or `SELECT * FROM Asset`? If you're doing the latter as your previous comment suggests, it will return all records from the DB, which is going to be slow. If you're just trying to get a count, there's no reason to do that.

**Comment by sacrampton** [comment]
Hi @iartemiev - here are a few points....

On iOS if I do the following query it gives me the correct answer and does so very quickly.    

SELECT COUNT(*) as count FROM Asset WHERE auditCompleted = true

If we execute exactly the same query on Android it gives this error

```
Object {
"code": 0,
"message": "Cannot bind argument at index 1 because the index is out of range. The statement has 0 parameters.",
}
```

We were using the other SELECT statement because we can't seem to get the COUNT to work with an AND statement - the following query in iOS returns an incorrect value of 1. In Android it returns the same error.

SELECT COUNT(*) as count FROM Asset WHERE auditCompleted = true AND packageName = JONESBORO

**Comment by iartemiev** [comment]
Try parameterizing the query and enclose the WHERE clause expression in parentheses. 

E.g., 
```js
await db.executeSql('SELECT COUNT(*) as count FROM Asset WHERE (auditCompleted = ? AND packageName = ?)', [true, 'JONESBORO']);
```

**Comment by sacrampton** [comment]
Hi @iartemiev - thanks for that - did the trick now. So resolved all issues. Thanks so much as always.

**Comment by iartemiev** [comment]
My pleasure! Very glad to hear that things are finally coming together!

> Still noticing the delta-sync on SQLite is slower than before. Not sure if you were seeing that.

Hmm, I remember you mentioning that in your email, but I haven't observed it yet myself. I'll do some more benchmarking around this specifically when I get a chance.
* How big is the performance difference? 
* Is the number of items returned by Delta sync similar when you were testing this with AsyncStorage vs. now with SQLite?

**Comment by mjaydeep01** [comment]
> Hi @iartemiev - some more updates here.
> 
> We deleted node modules and pod files, and follow all the steps above, but when we check the code it is still getting data from AsyncStorage.
> 
> ```
> const model = 'Asset';
>  
>  const key = (await AsyncStorage.getAllKeys()).find((key) =>
>  key.includes(`@AmplifyDatastore::user_${model}`),
>  );
>  
>  console.log('Item from aync storage >>', await AsyncStorage.getItem(key));
> ```
> 
> So clearly we are not doing something right here. Can you see if there is anything obvious that we are omitting.
> 
> Suspect something is being cached in existing project so will create a new project from scratch, but that will take several hours, an see if that works for us.

@iartemiev, I am facing same issue on Android, no matter what i do, data is always being stored in the AsyncStorage. Steps, I followed are as below..
1. Delete node_modules folder, any .lock file.
2. Remove app from the Android device.
3. npm install
4. npm install aws-amplify@rn-sqlite amazon-cognito-identity-js@rn-sqlite aws-amplify-react-native@rn-sqlite react-native-sqlite-storage
4. Install app on the device.
4. On querying the data from datastore, data downloaded from the appsync, but stored in the AsyncStorage.

Am I missing any specific setting?

**Comment by sacrampton** [comment]
Hi @iartemiev - want to push further into the slowness we are seeing for DeltaSync

DataStore creates a separate table in DynamoDB to manage the DeltaSync called "AmplifyDataStore-ENV".

There are no indexes in this table - just the partition key and sort key - where the partition key is table/date and sort key is time/id/version.

Our database is multi-tenanted - and we deal with assets in industrial plants. So I could have hundreds of other users in other plants making massive amounts of changes. But I might not have any users working my plant. The DeltaSync as I see it is going to have to sort through everyone else's changes just to work out there are zero changes that are going to be applicable to me.

When we initially hydrate the cache we do a base query which uses GSI's to get a quick response. 

At the moment I'm seeing DeltaSync take about the same amount of time as the full sync (20 minutes). Today I know I was doing a lot of bulk updating of data in a few different plants through our web back end. Not an unusually large workload. But I am concerned from the slowness I'm seeing in our database and what I see in the DeltaSync table for DynamoDB has me worried that this is not scalable for a multi-tenanted environment.

You've been really good at coming up with solutions to get us moving - hopefully someone else has already come up with a solution to make the DeltaSync run in seconds rather than 20+ minutes.

**Comment by iartemiev** [comment]
@mjaydeep01 try deleting your node_modules and lock file, then adding the following to your package.json and then running `npm install`:
```json
  "resolutions": {
    "@aws-amplify/datastore": "rn-sqlite",
    "@aws-amplify/core": "rn-sqlite",
    "@aws-amplify/*": "rn-sqlite"
  },
```

**Comment by iartemiev** [comment]
@sacrampton, I think this behavior likely warrants a separate GitHub issue, unless this is somehow related to the on-device database on React Native specifically (AsyncStorage or SQLite). 

To better understand what's going on, I have some follow up questions:

1. How many total records are in the delta sync table in DynamoDB at the time that you're seeing the 20 min delta sync time?
2. How many of those records are being synced down to the app?
3. Are you using `DataStore.configure` to change any of the sync-related settings (e.g., `syncPageSize`, `fullSyncInterval`, etc.)? If so, which settings are you using?
4. Are you seeing roughly the same delta sync performance if you test this in a web app?

**Comment by sacrampton** [comment]
Hi @iartemiev - have created a separate GitHub issue as requested - #8699

**Comment by mjaydeep01** [comment]
@iartemiev, I have it added.

I think, resolutions section is not supported with npm. I tried with "npm install", but nothing happened.
With "yarn install", I am getting following warnings, so seems yarn also ignoring the specified resolutions.
Does it have to some do something with yarn version? Or any hint?

yarn install v1.22.0
warning Resolution field "rn-sqlite" has an invalid version entry and may be ignored
warning Resolution field "rn-sqlite" has an invalid version entry and may be ignored
warning Resolution field "@aws-amplify/*" does not end with a valid package name and will be ignored
[1/4] Resolving packages...
success Already up-to-date.

**Comment by mjaydeep01** [comment]
@iartemiev, thanks for the new suggestion...
npm does not seems to using this resolutions at all.
With Yarn install, i am getting following errors..

warning Resolution field "rn-sqlite" has an invalid version entry and may be ignored
warning Resolution field "rn-sqlite" has an invalid version entry and may be ignored
warning Resolution field "@aws-amplify/*" does not end with a valid package name and will be ignored

yarn --version --> 1.22.5
npm --version --> 7.20.5

Can you please suggest how to resolve this issues?
Can you

**Comment by iartemiev** [comment]
@sacrampton, @jeremy-white, @mjaydeep01 - this feature has been officially released as part of `aws-amplify@4.2.6`! The documentation change will follow shortly, but in the meantime, here's how you can enable it in your app:

(Note: this new storage adapter is only compatible with React Native CLI-generated apps. Expo support will be added in the future)

Install:
```sh
$ npm install aws-amplify @aws-amplify/datastore-storage-adapter react-native-sqlite-storage aws-amplify-react-native amazon-cognito-identity-js @react-native-community/netinfo @react-native-async-storage/async-storage
$ npx pod-install
```

Enable in your app:
```js
import { DataStore } from 'aws-amplify';
import { SQLiteAdapter } from '@aws-amplify/datastore-storage-adapter';

DataStore.configure({
  storageAdapter: SQLiteAdapter
});
```

I'm closing this issue, as the new storage adapter addresses the performance concerns expressed in the issue. Please create a new issue if you require assistance with this feature.

**Comment by sacrampton** [comment]
Hi @iartemiev - thank you for going above and beyond to make this happen

**Comment by prasadk-foodhub** [comment]
> ```shell
> npm install aws-amplify @aws-amplify/datastore-storage-adapter react-native-sqlite-storage aws-amplify-react-native amazon-cognito-identity-js @react-native-community/netinfo @react-native-async-storage/async-storage
> ```

SqliteAdapter prints empty array.Please confirm for me sync is not running after doing this

**Comment by prasadk-foodhub** [comment]
no such table: Setting
no such table: ModelMetadata

This are the errors I can see on logs.Please advise .I tried below code

import { DataStore } from 'aws-amplify';
import { SQLiteAdapter } from '@aws-amplify/datastore-storage-adapter';

DataStore.configure({
  storageAdapter: SQLiteAdapter
});

**Comment by RemyNtshaykolo** [comment]
> no such table: Setting no such table: ModelMetadata
> 
> This are the errors logs.Please advise .I tried below code
> 
> import { DataStore } from 'aws-amplify'; import { SQLiteAdapter } from '@aws-amplify/datastore-storage-adapter';
> 
> DataStore.configure({ storageAdapter: SQLiteAdapter });

Hey have you found a solution for this issue?
